File: PicPay/app/src/main/java/br/com/everaldocardosodearaujo/picpay/Activity/TransactionActivity.java
package br.com.everaldocardosodearaujo.picpay.Activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.squareup.picasso.Picasso;

import org.w3c.dom.Text;

import br.com.everaldocardosodearaujo.picpay.App.FunctionsApp;
import br.com.everaldocardosodearaujo.picpay.R;
import de.hdodenhof.circleimageview.CircleImageView;

import static br.com.everaldocardosodearaujo.picpay.App.FunctionsApp.closeActivity;
import static br.com.everaldocardosodearaujo.picpay.App.FunctionsApp.formatMask;
import static br.com.everaldocardosodearaujo.picpay.App.FunctionsApp.startActivity;
import static br.com.everaldocardosodearaujo.picpay.App.SessionApp.MASK_NUMBER;
import static br.com.everaldocardosodearaujo.picpay.App.SessionApp.TB_CREDIT_CARD;

public class TransactionActivity extends Activity {

    private CircleImageView idImg;
    private TextView idUserName;
    private TextView idName;
    private Button  idBtnPay;
    private TextView idTxvCardNumber;
    private EditText idValue;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_transaction);
        this.inflate();
        this.loadParams();
    }

    private void inflate(){
        this.idImg = (CircleImageView) findViewById(R.id.idImg);
        this.idName = (TextView) findViewById(R.id.idName);
        this.idUserName = (TextView) findViewById(R.id.idUserName);
        this.idBtnPay = (Button) findViewById(R.id.idBtnPay);
        this.idTxvCardNumber = (TextView) findViewById(R.id.idTxvCardNumber);
        this.idValue = (EditText)  findViewById(R.id.idValue);
    }

    private void loadParams(){
        try{
            Intent intent = this.getIntent();
            this.idName.setText(intent.getExtras().getString("name"));
            this.idUserName.setText(intent.getExtras().getString("username"));
            if (!intent.getExtras().getString("img").equals("")){
                Picasso.with(TransactionActivity.this).load(intent.getExtras().getString("img")).into(this.idImg);
            }
            String numberCard = TB_CREDIT_CARD.select().get(0).getNumberCard();
            this.idTxvCardNumber.setText("Cartão de crédito final " + numberCard.substring(15,numberCard.length()));
        }catch (Exception ex){
            Toast.makeText(TransactionActivity.this,"Ocorreu um erro: " + ex.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    public void onClickPay(View view){
        try{
            Toast.makeText(TransactionActivity.this,"Em desenvolvimento", Toast.LENGTH_SHORT).show();
        }catch (Exception ex){
            Toast.makeText(TransactionActivity.this,"Ocorreu um erro: " + ex.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    public void onClickShowCreditCard(View view){
        FunctionsApp.startActivity(TransactionActivity.this,CreditCardActivity.class,null);
    }
    @Override
    public void onResume(){
        super.onResume();
        try{
            String numberCard = TB_CREDIT_CARD.select().get(0).getNumberCard();
            this.idTxvCardNumber.setText("Cartão de crédito final " + numberCard.substring(15,numberCard.length()));
        }catch (Exception ex){
            Toast.makeText(TransactionActivity.this,"Ocorreu um erro: " + ex.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }
    public void onClickReturn(View view){
        closeActivity(TransactionActivity.this);
    }
}

File: app/src/main/java/br/com/vdsantana/picpaytest/transaction/TransactionActivity.kt
package br.com.vdsantana.picpaytest.transaction

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.EditText
import android.widget.Toast
import br.com.vdsantana.picpaytest.BaseActivity
import br.com.vdsantana.picpaytest.R
import br.com.vdsantana.picpaytest.creditcard.CreditCard
import br.com.vdsantana.picpaytest.creditcard.choose.ChooseCreditCardActivity
import br.com.vdsantana.picpaytest.transaction.di.DaggerTransactionActivityComponent
import br.com.vdsantana.picpaytest.transaction.di.TransactionActivityModule
import br.com.vdsantana.picpaytest.transaction.presenter.TransactionPresenter
import br.com.vdsantana.picpaytest.transaction.presenter.TransactionView
import br.com.vdsantana.picpaytest.users.User
import br.com.vdsantana.picpaytest.utils.picasso.CircleTransformation
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_transaction.*
import javax.inject.Inject

class TransactionActivity : BaseActivity(), TransactionView {

    @Inject
    lateinit var mTransactionPresenter: TransactionPresenter

    private val EXTRA_USER = "extra_user"
    private var selectedCreditCard: CreditCard? = null
    private var SELECT_CREDIT_CARD_REQUEST: Int = 102
    private var user: User? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_transaction)
        initializeUI()
    }

    private fun initializeUI() {
        user = intent.getSerializableExtra(EXTRA_USER) as User
        Picasso.get()
                .load(user?.img)
                .placeholder(R.drawable.ic_default_avatar)
                .transform(CircleTransformation())
                .into(avatar);
        username.text = user?.username
    }

    fun clickChooseCreditCard(v: View?) {
        startActivityForResult(Intent(this, ChooseCreditCardActivity::class.java), SELECT_CREDIT_CARD_REQUEST)
    }

    fun clickPayTransaction(v: View?) {
        if (selectedCreditCard == null) {
            Toast.makeText(this, "É necessário escolher um cartão de crédito!", Toast.LENGTH_LONG).show()
            return
        }

        if (editAmount.text.length == 0 || editAmount.currencyDouble < 0.01) {
            Toast.makeText(this, "Digite um valor válido", Toast.LENGTH_LONG).show()
            return
        }

        showCvvDialog()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == SELECT_CREDIT_CARD_REQUEST && resultCode == Activity.RESULT_OK) {
            selectedCreditCard = data?.extras?.get("selected_card") as CreditCard
            txtFinalCard.text = getString(R.string.final_card, selectedCreditCard?.cardNumber!!.substring(selectedCreditCard?.cardNumber!!.length.minus(4)))
        }
    }

    override fun onTransactionRequestSuccess(transactionResponse: TransactionResponse) {
        if (transactionResponse.transaction.success) {
            Toast.makeText(this, "Transação realizada com sucesso!", Toast.LENGTH_LONG).show()
            finish()
        } else
            onError()
    }

    override fun showProgress() {
        progressTransaction.visibility = View.VISIBLE
        layForm.visibility = View.GONE
    }

    override fun onError() {
        Toast.makeText(this, "Falha ao realizar transação!", Toast.LENGTH_LONG).show()
        progressTransaction.visibility = View.GONE
        layForm.visibility = View.VISIBLE
    }

    override fun onActivityInject() {
        DaggerTransactionActivityComponent.builder().appComponent(getAppcomponent())
                .transactionActivityModule(TransactionActivityModule())
                .build()
                .inject(this)

        mTransactionPresenter.attachView(this)
    }

    fun showCvvDialog() {
        val dialogBuilder = AlertDialog.Builder(this)
        var cvvDialog: AlertDialog? = null
        val inflater = this.layoutInflater
        val dialogView = inflater.inflate(R.layout.dialog_cvv, null)
        dialogBuilder.setView(dialogView)

        val editCVV = dialogView.findViewById<View>(R.id.editCVV) as EditText

        dialogBuilder.setTitle("CVV")
        dialogBuilder.setMessage("Digite o CVV do seu cartão")

        dialogBuilder.setPositiveButton("Continuar", { dialog, whichButton ->
            if (!editCVV.text.isEmpty()) {
                mTransactionPresenter.sendTransaction(selectedCreditCard?.cardNumber, editCVV.text.toString().toInt(), editAmount.currencyDouble, selectedCreditCard?.expiryDate, user?.id)
                cvvDialog?.dismiss()
            }
        })

        dialogBuilder.setNegativeButton("Cancelar", { dialog, whichButton ->
            cvvDialog?.dismiss()
        })

        cvvDialog = dialogBuilder.create()
        cvvDialog.show()
    }
}
